{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, we will be taking a different approach to looking at extinction. Here, we'll rescue a group of individuals that are at a critical density and look to see if genetic diversity can be recovered within this population. \n",
    "\n",
    "To model this, we'll assign a linear vector of SNPs to each individual. Each element in the vector is 0, 1, or 2 which corresponds to the number of mutant polymorphisms in the genome at that position. For example, if we have a G -> A mutation, we would put 0 for GG, 1 for GA or AG, or 2 for AA. This linear vector is based on a pair of binary vectors of the same length, which determines whether that SNP is the mutant on a specific homolog. The 0, 1, 2 vector is obviously calculated then by element-wise adding the two vectors.\n",
    "\n",
    "Thus, a population can be modeled as a matrix. The model will randomly generate a matrix or take in an input matrix from the user. What it needs to take in, however, is either a distance array or an LD array. For $n$ SNPs, this means an $n-1$ length vector of either base pair distances or LD values between the SNPs. That is, the 0th element of the linkage array corresponds to the linkage between the 0th and 1st SNPs. \n",
    "\n",
    "New individuals are created by choosing two individuals. For each parent, we 50/50 start with one of the two homologs, then pick the next mutant based on the provided distance/linkage disequilibrium. This goes until the end of the vector. The two vectors are then combined to make the new individual.\n",
    "\n",
    "We assume no other mutations, and that all parents die after each reproduction period, so all new individuals in the next generation are good. We also assume linear growth from provided critical number to provided thriving population number, but a critical assumption here is that one new individual is created per generation. This reflects (often) low reproduction rates with many endangered species, especially larger animals and mammals. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_genetic_recovery(n_init,\n",
    "                              n_thriving,\n",
    "                              num_snps,\n",
    "                              ld_array = None,\n",
    "                              dist_array = None,\n",
    "                              snp_mats = None):\n",
    "    \"\"\"\n",
    "    Simulates a model of genetic recovery, where we start with a population of n_init and add one new individual\n",
    "    until we reach n_thriving. The new individual's genotype is determined with probabilities based on the \n",
    "    provided LD or distance array. \n",
    "\n",
    "    Inputs:\n",
    "    n_init: int, initial population size\n",
    "    n_thriving: int, target population size\n",
    "    num_snps: int, number of SNPs\n",
    "    ld_array: 1D array of length num_snps-1, optional. This array contains the pairwise LD values between the SNPs.\n",
    "    dist_array: 1D array of length num_snps-1, optional. This array contains the pairwise distances between the SNPs in megabases, or centimorgans.\n",
    "        Note that one of ld_array or dist_array must be provided, and if both are, ld_array will be used.\n",
    "    snp_mats: Array of 2D-arrays, optional. The number of 2D arrays must be equal to n_init. \n",
    "        Each 2D array must have shape (2, num_snps); each row is binary and contains the genotype of the individual on the paternal\n",
    "        or maternal chromosome. It doesn't particularly matter which is which, but the first row is assumed to be the paternal chromosome.\n",
    "        If this is not provided, the function will generate random genotypes for the initial population.\n",
    "\n",
    "    Returns:\n",
    "    snp_mats: Array of 2D-arrays. The number of 2D arrays is equal to n_thriving. \n",
    "        Each 2D array has shape (2, num_snps); each row is binary and contains the genotype of the individual on the paternal\n",
    "        or maternal chromosome. It doesn't particularly matter which is which, but the first row is assumed to be the paternal chromosome.\n",
    "    \"\"\"\n",
    "\n",
    "    #first: one of ld_array or dist_array must be provided\n",
    "    if ld_array is None and dist_array is None:\n",
    "        raise ValueError(\"Either ld_array or dist_array must be provided\")\n",
    "    \n",
    "    #second: if snp_mats is not provided, generate random genotypes for the initial population\n",
    "    if snp_mats is None:\n",
    "        snp_mats = generate_random_genos(n_init, num_snps)\n",
    "    elif len(snp_mats) != n_init:\n",
    "        raise ValueError(\"The number of provided SNP matrices must be equal to n_init\")\n",
    "    elif snp_mats[0].shape[1] != num_snps:\n",
    "        raise ValueError(\"The number of SNPs in the provided SNP matrices must be equal to num_snps\")\n",
    "    elif len(snp_mats[0]) != 2:\n",
    "        raise ValueError(\"Each SNP matrix must have shape (2, num_snps)\")\n",
    "    \n",
    "    probs = None\n",
    "    #third: if ld_array is provided, check that it has the correct length; if it does, convert to probabilities\n",
    "    if ld_array is not None:\n",
    "        if len(ld_array) != num_snps - 1:\n",
    "            raise ValueError(\"The length of ld_array must be equal to num_snps - 1\")\n",
    "        else:\n",
    "            print(\"TODO\")\n",
    "\n",
    "    #fourth: if dist_array is provided, check that it has the correct length; if it does, convert to probabilities\n",
    "    if dist_array is not None:\n",
    "        if len(dist_array) != num_snps - 1:\n",
    "            raise ValueError(\"The length of dist_array must be equal to num_snps - 1\")\n",
    "        else:\n",
    "            print(\"TODO\")\n",
    "\n",
    "    for i in range(n_init, n_thriving):\n",
    "        next_gen = []\n",
    "        for j in range(i + 1):\n",
    "            #pick two random individuals from the population\n",
    "            ind1 = np.random.randint(0, len(snp_mats))\n",
    "            ind2 = np.random.randint(0, len(snp_mats))\n",
    "            while ind1 == ind2:\n",
    "                ind2 = np.random.randint(0, len(snp_mats)) #ensure they are different\n",
    "        \n",
    "            #using these, reproduce a new individual\n",
    "            new_ind = generate_new_ind(snp_mats[ind1], snp_mats[ind2], probs)\n",
    "\n",
    "            #add the new individual to the population\n",
    "            next_gen.append(new_ind)\n",
    "\n",
    "        snp_mats = next_gen\n",
    "        #print the current population size\n",
    "        print(\"Current population size: \", len(snp_mats))\n",
    "\n",
    "    return None\n",
    "\n",
    "#helper function: generate random genotypes, given n_init and num_snps\n",
    "def generate_random_genos(n_init, num_snps):\n",
    "    snp_mats = []\n",
    "    for i in range(n_init):\n",
    "        snp_mat = np.random.randint(0, 2, size=(2, num_snps))\n",
    "        snp_mats.append(snp_mat)\n",
    "    return snp_mats\n",
    "\n",
    "#helper function: given an LD array, convert to probabilities\n",
    "#for simplicity: anything above 0.5 is kept as the original, and anything below is raised to 0.5\n",
    "def convert_ld_to_probs(ld_array):\n",
    "    probs = np.zeros(len(ld_array))\n",
    "    for i in range(len(ld_array)):\n",
    "        if ld_array[i] > 0.5:\n",
    "            probs[i] = ld_array[i]\n",
    "        else:\n",
    "            probs[i] = 0.5\n",
    "    return probs\n",
    "\n",
    "#helper function: given a distance array, convert to probabilities\n",
    "#for simplicity: anything distance x below 50 is converted to 1 - (x/100), and anything above is converted to 0.5\n",
    "def convert_dist_to_probs(dist_array):\n",
    "    probs = np.zeros(len(dist_array))\n",
    "    for i in range(len(dist_array)):\n",
    "        if dist_array[i] < 50:\n",
    "            probs[i] = 1 - (dist_array[i]/100)\n",
    "        else:\n",
    "            probs[i] = 0.5\n",
    "    return probs\n",
    "\n",
    "#helper function: given two individuals' SNP matrices and a probability array, generate a new individual\n",
    "def generate_new_ind(ind1, ind2, probs):\n",
    "    num_snps = ind1.shape[1]\n",
    "    new_ind = np.zeros((2, num_snps))\n",
    "\n",
    "    #first, the paternal chromosome from ind1\n",
    "    current_chrom = np.random.randint(0, 2) #0 or 1\n",
    "    for i in range(num_snps):\n",
    "        new_ind[0, i] = ind1[current_chrom, i]\n",
    "\n",
    "        #flip a coin to determine whether to use the other chromosome\n",
    "        if i == num_snps - 1:\n",
    "            continue\n",
    "\n",
    "        flip = np.random.rand()\n",
    "        if flip > probs[i]:\n",
    "            current_chrom = 1 - current_chrom\n",
    "\n",
    "    #second, the maternal chromosome from ind2\n",
    "    current_chrom = np.random.randint(0, 2) #0 or 1\n",
    "    for i in range(num_snps):\n",
    "        new_ind[1, i] = ind2[current_chrom, i]\n",
    "\n",
    "        #flip a coin to determine whether to use the other chromosome\n",
    "        if i == num_snps - 1:\n",
    "            continue\n",
    "\n",
    "        flip = np.random.rand()\n",
    "        if flip > probs[i]:\n",
    "            current_chrom = 1 - current_chrom\n",
    "\n",
    "    return new_ind\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([[0, 1, 0, 1, 1],\n",
       "        [1, 1, 1, 0, 1]]),\n",
       " array([[1, 1, 0, 1, 1],\n",
       "        [1, 1, 1, 0, 1]]),\n",
       " array([[1, 1, 0, 0, 1],\n",
       "        [0, 1, 1, 0, 0]]),\n",
       " array([[0, 0, 0, 0, 1],\n",
       "        [0, 1, 1, 1, 1]]),\n",
       " array([[1, 0, 0, 0, 1],\n",
       "        [0, 1, 0, 0, 0]]),\n",
       " array([[1, 1, 1, 1, 0],\n",
       "        [0, 0, 0, 1, 1]]),\n",
       " array([[0, 1, 1, 0, 1],\n",
       "        [0, 1, 1, 0, 1]]),\n",
       " array([[0, 0, 0, 0, 0],\n",
       "        [1, 1, 1, 1, 0]]),\n",
       " array([[0, 0, 0, 1, 0],\n",
       "        [0, 0, 0, 1, 1]]),\n",
       " array([[1, 0, 0, 0, 1],\n",
       "        [0, 0, 0, 0, 0]])]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_random_genos(10, 5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
